frontend/ts/courses.ts

```diff
@@ -0,0 +1,47 @@
+import { load as loadYAML } from 'js-yaml'
+import { getUserData } from './storage'
+
+interface Chapter {
+  title: string,
+  id: string,
+  url: string,
+  progress: number
+}
+
+interface Course {
+  title: string,
+  url: string,
+  id: string
+  type: string,
+  sections: Chapter[]
+}
+let courseList: Course[] | undefined
+
+const getCourseList = () : Course[] => {
+  if (!courseList) {
+    courseList = (loadYAML(document.getElementById('toc')?.textContent || '') || []) as [Course]
+    courseList.forEach((course) => {
+      course.id = course.url.startsWith('/') ? course.url.substring(1) : course.url
+      course.sections.forEach((section) => {
+        section.url = `/course/${course.id}/${section.id}`
+        section.progress = getUserData()?.[course.id]?.[section.id]?.progress
+      })
+    })
+  }
```

Comment:
from my understanding, `getUserData()` this call could be moved outside of the nested `forEach` statements so it is not called repeatedly. i.e.,

```suggestion
  if (!courseList) {
    const userData = getUserData()
    courseList = (loadYAML(document.getElementById('toc')?.textContent || '') || []) as [Course]
    courseList.forEach((course) => {
      course.id = course.url.startsWith('/') ? course.url.substring(1) : course.url
      course.sections.forEach((section) => {
        section.url = `/course/${course.id}/${section.id}`
        section.progress = userData?.[course.id]?.[section.id]?.progress
      })
    })
  }
```